README.md
# dChipIO: Methods for Reading dChip Files


## Installation
R package dChipIO is available on [CRAN](http://cran.r-project.org/package=dChipIO) and can be installed in R as:
```r
install.packages('dChipIO')
```




## Software status

| Resource:     | CRAN        | Travis CI     | Appveyor         |
| ------------- | ------------------- | ------------- | ---------------- |
| _Platforms:_  | _Multiple_          | _Linux_       | _Windows_        |
| R CMD check   | <a href="http://cran.r-project.org/web/checks/check_results_dChipIO.html"><img border="0" src="http://www.r-pkg.org/badges/version/dChipIO" alt="CRAN version"></a> | <a href="https://travis-ci.org/HenrikBengtsson/dChipIO"><img src="https://travis-ci.org/HenrikBengtsson/dChipIO.svg" alt="Build status"></a> | <a href="https://ci.appveyor.com/project/HenrikBengtsson/dchipio"><img src="https://ci.appveyor.com/api/projects/status/github/HenrikBengtsson/dChipIO?svg=true" alt="Build status"></a> |
| Test coverage |                     | <a href="https://coveralls.io/r/HenrikBengtsson/dChipIO"><img src="https://coveralls.io/repos/HenrikBengtsson/dChipIO/badge.svg?branch=develop" alt="Coverage Status"/></a>   |                  |


NEWS
Package: dChipIO
================

Version: 0.1.4-9000 [2016-01-12]
o BUG FIX: The read functions did not work on big-endian platform
  (e.g. Solaris SPARC), because they read files using the default
  platform endian rather than enforcing little endian.
o CLEANUP: Removed warning on readBin() and "'signed = FALSE' is
  only valid for integers of sizes 1 and 2".


Version: 0.1.4 [2016-01-12]
o CLEANUP: Previous version had a stray Rplots.pdf in tests/.


Version: 0.1.3 [2016-01-12]
o CRAN Policies: Package Title Case and proper starting of description.
o ROBUSTNESS: Add package tests (from examples). Test coverage is %97.
o Updated broken Affymetrix URL.


Version: 0.1.2 [2014-06-16]
o Fixed 'R CMD check' notes appearing in more recent versions of R.


Version: 0.1.1 [2009-02-13]
o CLEAN UP: readCdfBinHeader() and readDcpHeader() no longer give
  warnings on "In rawToChar(...) : truncating string with embedded nul".


Version: 0.1.0 [2008-08-20]
o Added example data.
o Added readDcpRectangle().
o The readDcp() and readDcpHeader() was taken from the
  aroma.affymetrix package.   The readCdfBin() and
  readCdfBinHeader() was taken from the authors private
  set of code.  The first versions of these functions
  were written on January 30, 2008.
o Package created.


R/utils.R
.rawToString <- function(raw, ...) {
  # This approach drops all '\0', in order to avoid warnings
  # in rawToChar().  Note, it does not truncate the string after
  # the first '\0'.  However, such strings should never occur in
  # the first place.
  raw <- raw[raw != as.raw(0)]
  rawToChar(raw)
} # .rawToString()

.readRaw <- function(con, n=1L, ...) {
  readBin(con, what="raw", n=n)
}

.readByte <- function(con, n=1L, ...) {
  readBin(con, what="integer", n=n, size=1L, signed=FALSE)
}

.readUShort <- function(con, n=1L, ...) {
  readBin(con, what="integer", n=n, size=2L, signed=FALSE, endian="little")
}

.readInt <- function(con, n=1L, ...) {
  readBin(con, what="integer", n=n, size=4L, signed=TRUE, endian="little")
}

.readFloat <- function(con, n=1L, ...) {
  readBin(con=con, what="double", n=n, size=4L, signed=TRUE, endian="little")
}

.readString <- function(con, n, ...) {
  .rawToString(.readRaw(con, n=n))
}


R/readCdfBin.R
###########################################################################/**
# @RdocFunction readCdfBin
#
# @title "Reads a dChip CDF.bin file"
#
# \description{
#   @get "title".
#
#   Please note that this method is incomplete as it currently doesn't read
#   all fields.  It is only made available so that someelse can continue
#   the development.
# }
#
# @synopsis
#
# \arguments{
#   \item{con}{A @connection or a @character filename.}
#   \item{units}{An @integer @vector specifying the units to be read.
#     If @NULL, all units are read.}
#   \item{...}{Not used.}
# }
#
# \value{
#   Returns a @list structure containing the file header and the unit data.
# }
#
# @examples "../incl/readCdfBin.Rex"
#
# @author
#
# \seealso{
#   To read only the CDF.bin file header, see @see "readCdfBinHeader".
# }
#
# @keyword "file"
# @keyword "IO"
#*/###########################################################################
readCdfBin <- function(con, units=NULL, ...) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Validate arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Argument 'con':
  if (is.character(con)) {
    pathname <- con;
    if (!file.exists(pathname)) {
      stop("File not found: ", pathname);
    }
    con <- file(con, open="rb");
    on.exit({
      if (!is.null(con))
        close(con);
      con <- NULL;
    });
  }

  if (!inherits(con, "connection")) {
    stop("Argument 'con' must be either a connection or a pathname: ",
                                                            mode(con));
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Read CDF.bin header
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  hdr <- readCdfBinHeader(con=con);
  nbrOfUnits <- hdr$NumUnit;
  # Sanity checks
  stopifnot(nbrOfUnits >= 0)

  # Arguments 'units':
  if (is.null(units)) {
    units <- 1:nbrOfUnits;
  } else {
    units <- .argAssertRange(as.integer(units), range=c(1, nbrOfUnits));
    nbrOfUnits <- length(units);
  }

  data <- list(header=hdr);


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Read unit data
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  UNIT_NAME_LEN <- 50;
#	char UnitName[UNIT_NAME_LEN]; // as in *.cdf file
#	short NumProbe; // number of probe pairs
#	SHORT_POINT_ARRAY *CellPos;
  CDF_UNIT_SIZE = UNIT_NAME_LEN + 2 + 4;

  firstUnit <- min(units);
  lastUnit <- max(units);
  nbrOfUnitsToRead <- as.integer(lastUnit - firstUnit + 1);

  # Skip to the first unit to be read
  if (firstUnit > 0) {
    seek(con=con, origin="current", where=(firstUnit-1)*CDF_UNIT_SIZE,
                                                            rw="read");
  }

  # Read first to last unit...
  raw <- .readRaw(con, n=nbrOfUnitsToRead*CDF_UNIT_SIZE)


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Keep only units of interest
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  dim(raw) <- c(CDF_UNIT_SIZE, nbrOfUnitsToRead);
  keep <- match(units, (firstUnit:lastUnit));
  raw <- raw[,keep,drop=FALSE];

  # Extract 'UnitName'
  idxs <- 1:UNIT_NAME_LEN;
  unitNames <- rep("", nbrOfUnits);
  for (idx in idxs) {
    unitNames <- paste(unitNames,
                   rawToChar(raw[idx,,drop=TRUE], multiple=TRUE), sep="");
  }
  raw <- raw[-idxs,,drop=FALSE];
  data$unitNames <- unitNames;
  rm(unitNames);

  # Extract 'NumProbe'
  idxs <- 1:2;
  data$numProbes <- .readUShort(raw[idxs,,drop=FALSE], n=nbrOfUnits)
  raw <- raw[-idxs,,drop=FALSE];
  # Sanity checks
  stopifnot(data$numProbes >= 0)

  data$CellPos <- .readInt(raw, n=nbrOfUnits)
  # Sanity checks
  stopifnot(data$CellPos >= 0)

  rm(raw);


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Read cell data
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Skip to the last unit
  if (lastUnit < nbrOfUnits) {
    seek(con=con, origin="current",
                where=(nbrOfUnits-lastUnit)*CDF_UNIT_SIZE, rw="read");
  }

  # To do

  data;
}


##############################################################################
# HISTORY:
# 2008-08-20
# o BUG FIX: The wrong subset of units was read when 'units' was specified.
# o Added Rdoc comments.
# 2008-02-03
# o Added reader for CDF.bin unit data. Cell data are still to be done.
# o Created.
##############################################################################


R/999.package.R
#########################################################################/**
# @RdocPackage dChipIO
#
# \description{
#   @eval "packageDescription('dChipIO')$Description"
#
#   The example data used in this package orginates from the
#   Affymetrix Fusion SDK library [3].
# }
#
# \author{
#   Henrik Bengtsson.
# }
#
# \section{To get started}{
#   To get started, see:
#   \enumerate{
#     \item \code{\link{readCdfBin}}() - reads a dChip CDF.bin file.
#     \item \code{\link{readDcp}}() - reads a dChip DCP data file.
#   }
# }
#
# \section{License}{
#   The releases of this package is licensed under
#   LGPL version 2.1 or newer.
# }
#
# \references{
#   [1] The dChip software, \url{http://www.dchip.org/}\cr
#   [2] Thread 'DCP File Format', dChip forum 'Welcome, Download \& Codes',
#       2008. \url{http://dchip.forum5.com/viewtopic.php?p=812#812}
#       [This old forum is no longer available / Jan 2016]\cr
#   [3] Affymetrix Inc, Fusion Software Developers Kit (SDK), 2008.
#       \url{http://www.affymetrix.com/estore/partners_programs/programs/developer/fusion/index.affx?terms=no}
# }
#*/#########################################################################


R/readDcpHeader.R
###########################################################################/**
# @RdocFunction readDcpHeader
#
# @title "Reads the file header of a dChip DCP file"
#
# \description{
#   @get "title".
# }
#
# @synopsis
#
# \arguments{
#   \item{con}{A @connection or a @character filename.}
#   \item{...}{Not used.}
# }
#
# \value{
#   Returns a @list structure containing the file header.
# }
#
# @author
#
# \seealso{
#   To read also the DCP file data, see @see "readDcp".
# }
#
# @keyword "file"
# @keyword "IO"
#*/###########################################################################
readDcpHeader <- function(con, ...) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Validate arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Argument 'con':
  if (is.character(con)) {
    pathname <- con;
    if (!file.exists(pathname)) {
      stop("File not found: ", pathname);
    }
    con <- file(con, open="rb");
    on.exit({
      if (!is.null(con))
        close(con);
      con <- NULL;
    });
  }

  if (!inherits(con, "connection")) {
    stop("Argument 'con' must be either a connection or a pathname: ",
                                                            mode(con));
  }



  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Read the file header
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  hdr <- list();

  hdr$Header <- .readString(con, n=1000)
  hdr$Format <- .readByte(con, n=1)

  # Assert that the file format is correct/the expected on
  if (!is.element(hdr$Format, 3:4)) {
    stop("File format error: The DCP header format ('Format') is not v3 or v4: ", hdr$Format);
  }

  hdr$Normalized <- as.logical(.readByte(con, n=1))
  hdr$ThetaValid <- as.logical(.readByte(con, n=1))

  # Don't know why we have to add this one
  hdr$dummy <- .readRaw(con, n=1)
# print(seek(con, read="r"));

  hdr$Median <- .readInt(con, n=1)
  hdr$MaxInten <- .readInt(con, n=1)
  hdr$CellDim <- .readInt(con, n=1)
# print(seek(con, read="r"));

  hdr$DatFile <- .readString(con, n=1000)
# print(seek(con, read="r"));
  hdr$BaselineFile <- .readString(con, n=1000)
# print(seek(con, read="r"));

  hdr$ArrayOutlierPct <- .readFloat(con, n=1)
  hdr$SingleOutlierPct <- .readFloat(con, n=1)
  hdr$PresencePct <- .readFloat(con, n=1)
# print(seek(con, read="r"));


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Clean up the header
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  hdr$dummy <- NULL;
  hdr$Header <- hdr$Header
  hdr$DatFile <- hdr$DatFile
  hdr$BaselineFile <- hdr$BaselineFile

  hdr;
} # readDcpHeader()


##############################################################################
# HISTORY:
# 2009-02-13
# o Using .rawToString() instead of rawToChar() to avoid warnings on
#   'truncating string with embedded nul:...'.
# 2008-xx-xx
# o Added Rdoc comments.
# o Added argument 'clean'.
# 2008-01-30
# o Created.
##############################################################################



R/readCdfBinHeader.R
###########################################################################/**
# @RdocFunction readCdfBinHeader
#
# @title "Reads the file header of a dChip CDF.bin file"
#
# \description{
#   @get "title".
# }
#
# @synopsis
#
# \arguments{
#   \item{con}{A @connection or a @character filename.}
#   \item{...}{Not used.}
# }
#
# \value{
#   Returns a @list structure containing the file header.
# }
#
# @author
#
# \seealso{
#   To read the CDF.bin file data, see @see "readCdfBin".
# }
#
# @keyword "file"
# @keyword "IO"
#*/###########################################################################
readCdfBinHeader <- function(con, ...) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Validate arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Argument 'con':
  if (is.character(con)) {
    pathname <- con;
    if (!file.exists(pathname)) {
      stop("File not found: ", pathname);
    }
    con <- file(con, open="rb");
    on.exit({
      if (!is.null(con))
        close(con);
      con <- NULL;
    });
  }

  if (!inherits(con, "connection")) {
    stop("Argument 'con' must be either a connection or a pathname: ",
                                                            mode(con));
  }

  hdr <- list();

  LINE_LEN <- 1000;
  UNIT_NAME_LEN <- 50;
  hdr$FileName <- .readString(con, n=LINE_LEN-1)
  hdr$Format <- .readByte(con)
  hdr$ChipType <- .readString(con, n=UNIT_NAME_LEN)

  # Don't know why we have to add this, but in the dChip source code file
  # 'cdf.cpp' there is the following comment:
  # "// 11/11/05, +2 since allocation is in the unit of 4 bytes"
  hdr$dummy <- .readRaw(con, n=2)

  hdr$CellDim <- .readInt(con)
  hdr$NumUnit <- .readInt(con)


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Clean up the header
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  hdr$dummy <- NULL;

  hdr;
} # readCdfBinHeader()



##############################################################################
# HISTORY:
# 2009-02-13
# o Using .rawToString() instead of rawToChar() to avoid warnings on
#   'truncating string with embedded nul:...'.
# 2008-08-20
# o Added Rdoc comments.
# 2008-02-03
# o Created.
##############################################################################


R/readDcp.R
###########################################################################/**
# @RdocFunction readDcp
#
# @title "Reads a dChip DCP file"
#
# \description{
#   @get "title".
# }
#
# @synopsis
#
# \arguments{
#   \item{con}{A @connection or a @character filename.}
#   \item{fields}{A @character @vector specifying the fields to be read.}
#   \item{cells}{An @integer @vector specifying the indices of the cell
#     data to be read.}
#   \item{units}{An @integer @vector specifying the indices of the unit
#     data to be read.}
#   \item{.nbrOfUnits}{A @integer specifying the number of units available
#     in the file. If @NULL, this is inferred from the file size and the
#     file header. The dChip software itself instead uses the corrsponding
#     value in the CDF.bin file, but that file is specified by the user
#     leaving room for errors.}
#   \item{...}{Not used.}
# }
#
# \value{
#   Returns a @list structure containing the file header and the
#   requested data fields.
# }
#
# @examples "../incl/readDcp.Rex"
#
# @author
#
# \seealso{
#   To read only the DCP file header, see @see "readDcpHeader".
# }
#
# @keyword "file"
# @keyword "IO"
#*/###########################################################################
readDcp <- function(con, fields=c("rawIntensities", "normalizedIntensities", "calls", "thetas", "thetaStds", "excludes"),  cells=NULL, units=NULL, .nbrOfUnits=NULL, ...) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Local functions
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  readElements <- function(con, idxs, nbrOfElements, size=1, skip=FALSE, ..., drop=TRUE) {
    # Local functions
    readBin2 <- function(con, what, size, signed, n) {
      if (mode(what) == "raw") {
        n <- n * size;
        size <- 1;
      }
#print(list(what=what, size=size, signed=signed, n=n));
      readBin(con, what=what, size=size, signed=signed, n=n, endian="little")
    } # readBin2()


    if (skip) {
      seek(con, where=size*nbrOfElements, origin="current", rw="read")
      return(NULL);
    }

    # Read all elements
    values <- readBin2(con=con, size=size, ..., n=nbrOfElements);

    # Turn into a matrix?
    if (mode(values) == "raw") {
      # Sanity check
      if (length(values) %% size != 0) {
        stop("File format error/read error: The number of bytes read is not a multiple of ", size, ": ", length(values));
      }

      dim(values) <- c(size, (length(values) %/% size));

      # Keep only the subset of elements?
      if (!is.null(idxs)) {
        values <- values[,idxs,drop=FALSE];
      }

      # Drop singleton dimensions?
      if (drop)
        values <- drop(values);
    } else {
      if (!is.null(idxs))
        values <- values[idxs];
    }

    values;
  } # readElements()


  readCells <- function(con, cells, what=integer(), size=2, signed=FALSE, ...) {
    readElements(con=con, idxs=cells, nbrOfElements=nbrOfCells,
                                   what=what, size=size, signed=signed, ...);
  }


  readUnits <- function(con, units, what=double(), size=2, signed=FALSE, ...) {
    readElements(con=con, idxs=units, nbrOfElements=nbrOfUnits,
                                   what=what, size=size, signed=signed, ...);
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Validate arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Argument 'con':
  if (is.character(con)) {
    pathname <- con;
    if (!file.exists(pathname)) {
      stop("File not found: ", pathname);
    }

    con <- file(con, open="rb");
    on.exit({
      if (!is.null(con))
        close(con);
      con <- NULL;
    });
  }

  if (!inherits(con, "connection")) {
    stop("Argument 'con' must be either a connection or a pathname: ",
                                                            mode(con));
  }

  # Argument 'fields':
  fields <- match.arg(fields, several.ok=TRUE);


  # Reading file header
  res <- list();
  res$header <- readDcpHeader(con=con);
  nbrOfCells <- res$header$CellDim^2;
  # Sanity checks
  stopifnot(nbrOfCells >= 0)

  # Argument '.nbrOfUnits':
  if (is.null(.nbrOfUnits)) {
    # Inferring number of units from the file size.
    conInfo <- summary(con);
    if (conInfo$class != "file") {
      stop("Cannot infer the value of '.nbrOfUnits' from the connection, because it is not a file: ", conInfo$class);
    }

    pathname <- conInfo$description;
    if (!is.element(res$header$Format, c(3,4))) {
      stop("Cannot infer the value of '.nbrOfUnits' from the file. The file format is not v3 or v4 but v", res$header$Format, ": ", pathname);
    }

    nbrOfBytes <- file.info(pathname)$size;
    fileHeaderSize <- 3028;
    nbrOfUnitBytes <- nbrOfBytes - fileHeaderSize - 2*2*nbrOfCells;
    .nbrOfUnits <- as.integer(nbrOfUnitBytes / 13);

    # Sanity check
    if (nbrOfUnitBytes %% 13 != 0) {
      stop("Internal file format assumption error: Cannot infer the value of '.nbrOfUnits' from the file. The number of inferred bytes storing unit data is not a multiple of 13: ", nbrOfUnitBytes);
    }
  } else {
    .nbrOfUnits <- .argAssertRange(as.integer(.nbrOfUnits),
                                                    range=c(1, nbrOfCells));
    if (length(.nbrOfUnits) != 1)
      stop("Argument '.nbrOfUnits' must be a single integer.");
  }
  nbrOfUnits <- .nbrOfUnits;

  # Argument 'cells':
  if (!is.null(cells)) {
    cells <- .argAssertRange(as.integer(cells), range=c(1, nbrOfCells));
  }

  # Argument 'units':
  if (!is.null(units)) {
    units <- .argAssertRange(as.integer(units), range=c(1, nbrOfCells));
  }



  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Reading data section
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Reading/skipping raw and normalized intensities
  for (field in c("rawIntensities", "normalizedIntensities")) {
    res[[field]] <- readCells(con, cells=cells, skip=(!field %in% fields), drop=TRUE);
#    str(res);
  }

  # Reading/skipping calls
  field <- "calls";
  res[[field]] <- readUnits(con, units=units, what=raw(), size=1,
                                       skip=(!field %in% fields), drop=TRUE);
#  str(res);

  # Reading/skipping probe summaries
  skip <- !any(c("thetas", "thetaStds", "excludes") %in% fields);
  raw <- readUnits(con, units=units, what=raw(), size=12, skip=skip);

  # Parse probe summaries
  if (!skip) {
    n <- ncol(raw);

    field <- "thetas";
    if (field %in% fields) {
      res[[field]] <- .readFloat(con=raw[1:4,], n=n);
    }
    raw <- raw[-(1:4),,drop=FALSE];

    field <- "thetaStds";
    if (field %in% fields) {
      res[[field]] <- .readFloat(con=raw[1:4,], n=n);
    }
    raw <- raw[-(1:4),,drop=FALSE];

    field <- "excludes";
    if (field %in% fields) {
      res[[field]] <- .readInt(raw[1:4,], n=n);
    }
  }
  rm(raw);

# print(seek(con, read="r"));
#print(file.info(pathname2)$size);

  res;
} # readDcp()


##############################################################################
# HISTORY:
# 2008-08-20
# o BUG FIX: The wrong subset of units was read when 'units' was specified.
# o Added Rdoc comments.
# o If '.nbrOfUnits=NULL', then it is inferred automatically from the file
#   size, if possible.
# 2008-01-30
# o Created.
##############################################################################
